poseidon-sponge.md

```diff
@@ -0,0 +1,240 @@
+> `capacity || rate` instead, and at least one instance of Poseidon Merkle tree hashing
+> matching Appendix I was deployed into production before this inconsistency was noticed.
+>
+> TODO: Examine all open-source production instances of Poseidon to figure out how far
+> this inconsistency propagated, and what other variants might currently exist.
+
+[BDPV2007]: https://keccak.team/files/SpongeFunctions.pdf
+
+### Initialization
+
+`PoseidonSponge.Initialize` takes as input a `PoseidonPermutation`, a field element
+`domain_iv` that is used for domain separation, and a `duplex_padding` function. It
+proceeds as follows:
+
+1. Initialize the sponge state. The first element of the capacity portion is set to
+   `domain_iv`, while the remaining capacity elements, and all of the rate elements, are
+   initialized to zero.
+   ```
+   state = [0; T]
+   state[RATE] = domain_iv
+   ```
+
+2. Return the new sponge:
+   ```
+   PoseidonSponge(
+     PoseidonPermutation,
+     state,
+     mode = absorbing,
+     cache = [],
+     duplex_padding,
+   )
+   ```
+
+> TODO: Should `RATE` be passed directly to `PoseidonSponge.Initialize` instead of being
+> an implicit parameter?
+
+### Internal helper functions
+
+`ZeroExtend` is a function that takes an input of between zero and `RATE` field elements.
+It appends the array `[0; RATE - len(input)]` and returns the result.
+
+`PoseidonSponge.PerformDuplex` takes an input of between zero and `RATE` field elements.
+It proceeds as follows:
+
+1. Apply the domain's padding function to the input, resulting in a padded input of `RATE`
+   field elements.
+
+2. Add the padded input element-wise to the rate portion of `state`:
+   ```
+   for i in [0, RATE):
+       state[i] += padded_input[i]
+   ```
+
+3. Call `PoseidonPermutation(state)`.
+
+4. Return the first `RATE` field elements of `state` as output.
+
+### Absorb
+
+`PoseidonSponge.Absorb` takes a single field element `input`, and proceeds as follows:
+
+- If `mode == absorbing`:
+  - If `len(cache) == RATE`:
+    1. Call `PoseidonSponge.PerformDuplex(cache)`, and drop the result.
+    2. Set `cache` to `[input; 1]`.
+
+  - Else (there is room in `cache`):
+    1. Append `input` to the end of `cache`.
+
+- Else (`mode == squeezing`):
+  1. Drop any remaining cached output elements (by setting `cache` to `[]`).
+  2. Set `mode` to `absorbing`.
+  3. Set `cache` to `[input; 1]`.
+
+### Squeeze
+
+`PoseidonSponge.Squeeze` proceeds as follows:
+
+1. If `mode == squeezing` and `len(cache) == 0`:
+   1. Set `mode` to `absorbing`.
+   2. Set `cache` to `[]` (this is already the case, but for clarity this means there are
+      no elements to be absorbed).
+
+2. If `mode == absorbing`:
+   1. `new_output_elements = PoseidonSponge.PerformDuplex(cache)`.
+   2. Set `mode` to `squeezing`.
+   3. Set `cache` to `new_output_elements`.
+
+3. Remove the first element from `cache` and return it.
+
```

Comment:
The definition of `Squeeze` is also weird because it doesn't always return a value, but if you call it twice it will (I'm aware this is how Zcash does it). I think, for simplicity, `Squeeze` should always return a value.

Secondly, I think there's a bug in this definition. Say you've squeezed 1 element from the current state, `v = state[0]`. Then you absorb 1 element `w`. This sets `cache = [w]`. Finally, you squeeze once more. In the last squeeze, the XORed state has `state[0] = v + w`, i.e you've XORed a squeezed and absorbed value with each other. I didn't see it written anywhere, but this seems like something that should never happen.

I think the definition of these functions is a bit easier with a `frontier` index instead of a cache.
```
// Invariant: frontier is always in the range [0, RATE)

def Absorb(input):
    mode := ABSORBING

    for elem in input:
        state[frontier] += elem
        frontier += 1

        if frontier == RATE:
            state := permute(state)
            frontier := 0

def Squeeze(l):
    if self.mode == ABSORBING:
        // buf might of length RATE or 2*RATE depending on frontier. Absorb the
        // second half if it exists.
        buf := state || duplex_padding(frontier)
        state := permute(buf[..RATE])
        frontier := 0
        Absorb(buf[RATE..])

    output := state[frontier]

    frontier += 1
    if frontier == RATE:
        state := permute(state)
        frontier := 0

    return output
```
This doesn't have the same bug because the `frontier` keeps track of the last element touched.

Reply:
> The definition of Squeeze is also weird because it doesn't always return a value, but if you call it twice it will (I'm aware this is how Zcash does it). I think, for simplicity, Squeeze should always return a value.

I think you got tripped by the fact that this is a list of steps, not a either do 1 or 2. Which is great as another of my comment is that this might confuse implementers :p

Reply:
>  I didn't see it written anywhere, but this seems like something that should never happen.

It tripped me as well, but this is how the duplex function is defined

Reply:
> I think the definition of these functions is a bit easier with a frontier index instead of a cache.

I agree, but I think most implementations will use the `cache` method. If the two approaches are similar in the spec, but one is closer to the implementations, then that one should probably win.

Reply:
>  what does this allow you to do? You can reset to zero the rate part, but that won't help you much in finding a collision as you don't control the capacity.

You're right, this ends up being about as secure as having an `overwrite` operation (which exists in STROBE) that just writes to the state. I think the part that trips me up though is that two distinct sequences of absorbs and squeezes can lead to the same state. But I suppose that's OK for this primitive.

So yeah, no qualms about the cache method anymore, especially because its functionality differs from the thing that I wrote.

Reply:
> Secondly, I think there's a bug in this definition. Say you've squeezed 1 element from the current state, v = state[0]. Then you absorb 1 element w. This sets cache = [w]. Finally, you squeeze once more. In the last squeeze, the XORed state has state[0] = v + w, i.e you've XORed a squeezed and absorbed value with each other. I didn't see it written anywhere, but this seems like something that should never happen.

I presume this was due to the confusing syntax, but as I understand the duplex paper, there is no bug here. XORing the first-squeezed and absorbed value is exactly what would happen in Algorithm 2, prior to the permutation function being called (so the second-squeezed value is correctly bound to all prior absorbed values).